### PR TITLE
Add ExternalIP observation functions. 

### PR DESCRIPTION
**Pull request by squeed**
Used by kube-apsierver-operator and openshift-controller-manager-operator.

### CODE REVIEW

**Comment by squeed** [comment]
used by:

https://github.com/openshift/cluster-openshift-controller-manager-operator/pull/104
https://github.com/openshift/cluster-kube-apiserver-operator/pull/509

File: pkg/operator/configobserver/network/observe_network.go [hunk 1]
```diff
@@ -14,46 +15,116 @@ import (
 func GetClusterCIDRs(lister configlistersv1.NetworkLister, recorder events.Recorder) ([]string, error) {
 	network, err := lister.Get("cluster")
 	if errors.IsNotFound(err) {
-		recorder.Warningf("ObserveRestrictedCIDRFailed", "Required networks.%s/cluster not found", configv1.GroupName)
+		recorder.Warningf("GetClusterCIDRs", "Required networks.%s/cluster not found", configv1.GroupName)
```
Comment (danwinship) [inline pkg/operator/configobserver/network/observe_network.go]: You're inconsistent about "`GetClusterCIDRs`" vs "`GetClusterCIDRsFailed`" (and likewise in the other functions).

File: pkg/operator/configobserver/network/observe_network.go [hunk 2]
```diff
@@ -14,46 +15,116 @@ import (
...
+	}
+
+	if network.Spec.ExternalIP == nil {
+		return nil, nil
+	}
+
+	pol := network.Spec.ExternalIP.Policy
+	if pol != nil {
+		if err := validateCIDRs(pol.AllowedCIDRs); err != nil {
+			recorder.Warningf("GetExternalIPPolicy", "error parsing networks.%s/cluster Spec.ExternalIP.Policy.AllowedCIDRs: invalid cidr: %v", configv1.GroupName, err)
```
Comment (danwinship) [inline pkg/operator/configobserver/network/observe_network.go]: Could this be validated by the CRD?

File: pkg/operator/configobserver/network/observe_network.go [hunk 3]
```diff
@@ -14,46 +15,116 @@ import (
...
+			return nil, err
+		}
+	}
+
+	return network.Spec.ExternalIP.Policy, nil
+}
+
+// GetExternalIPPolicy retrieves the ExternalIPPolicy for the cluster.
+// The policy may be null.
+func GetExternalIPAutoAssignCIDRs(lister configlistersv1.NetworkLister, recorder events.Recorder) ([]string, error) {
```
Comment (danwinship) [inline pkg/operator/configobserver/network/observe_network.go]: comment is incorrect

File: pkg/operator/configobserver/network/observe_network.go (hunk 1 above)
Comment (squeed) [reply pkg/operator/configobserver/network/observe_network.go]: I was trying to match style (that's a Reason, not an event category or source, but there's no clear distinction.

File: pkg/operator/configobserver/network/observe_network.go (hunk 2 above)
Comment (squeed) [reply pkg/operator/configobserver/network/observe_network.go]: Maybe, yeah. Where the heck does the CRD actually live?

**Comment by squeed** [comment]
API change has merged, updated vendor.

If anyone knows how to get glide to do minimal updates, let me know. (I think the answer is to switch to `dep`)

File: pkg/operator/configobserver/network/observe_network.go (hunk 2 above)
Comment (squeed) [reply pkg/operator/configobserver/network/observe_network.go]: added a subtask to add validation to the cluster-config-operator.

**Comment by squeed** [comment]
This is ready for final review and merge. @danwinship @deads2k can you review, please?

**Comment by squeed** [comment]
/assign @deads2k 

Can you take a look?

**Comment by squeed** [comment]
@smarterclayton @deads2k @sttts this PR has a vendor bump, so only one of you can approve it.

PTAL - this holds up 2 more PRs

File: pkg/operator/configobserver/network/observe_network.go [hunk 4]
```diff
@@ -14,46 +15,115 @@ import (
...
+
 	return network.Status.ServiceNetwork[0], nil
 }
+
+// GetExternalIPPolicy retrieves the ExternalIPPolicy for the cluster.
+// The policy may be null.
+func GetExternalIPPolicy(lister configlistersv1.NetworkLister, recorder events.Recorder) (*configv1.ExternalIPPolicy, error) {
+	network, err := lister.Get("cluster")
+	if errors.IsNotFound(err) {
+		recorder.Warningf("GetExternalIPPolicyFailed", "Required networks.%s/cluster not found", configv1.GroupName)
```
Comment (sttts) [inline pkg/operator/configobserver/network/observe_network.go]: how much required is this? This will only lead to an event, not to any kind of error.

Comment (squeed) [reply pkg/operator/configobserver/network/observe_network.go]: I was just duplicating the style from previous functions. I can return nil

**Pull request merged**